draft-normington-jsonpath-latest.xml

```diff
@@ -381,57 +381,66 @@ array-slice = [ integer ] ws %x3A ws [ integer ]
-                        It selects no nodes from a node which is not an array.
-                     </t>
-                     <t>The members of a slice depend on whether the step value is positive, negative, or zero as follows. If the step value is
-                        omitted, it defaults to <tt>1</tt>.
-                     </t>
-                     <t>For positive integer <tt>step</tt> and integers <tt>start</tt> and <tt>end</tt>, the array slice
-                        <tt>start:end:step</tt> consists of the indices starting at <tt>start</tt> and increasing repeatedly
-                        by <tt>step</tt> while the resultant index is less than <tt>end</tt>.
-                        The indices of the slice are the same as the values of i iterated over by this <tt>for</tt> loop in C:
-                     </t>
+                     <t>A slice expression produces a new array containing a subset of the elements of the input array.</t>
+                     <t>A slice is defined by the two bounds, <tt>start</tt> and <tt>end</tt>,
+                        and a <tt>step</tt> which governs the direction of iteration and the iteration stride.
+                        Each of those parameters is optional.</t>
+                     <t>Negative bounds are interpreted as offsets from the end of the array, namely <tt>b + len</tt>.
+                        This remapping is performed before any other evaluation step.</t>
+                     <t>A slice bound <tt>b</tt> is considered valid if <tt>0 &lt;= b &lt; len</tt>.
+                        An invalid bound is treated as an omitted bound. Omitted bounds take a default value
+                        that depends on the value of the <tt>step</tt> parameter:</t>
+                     <t>If <tt>step &gt; 0</tt>, then <tt>start</tt> defaults to <tt>0</tt>
+                        while <tt>end</tt> defaults to <tt>len</tt>.</t>
+                     <t>If <tt>step &lt; 0</tt>, then <tt>start</tt> defaults to <tt>len-1</tt>,
+                        while an omitted <tt>end</tt> causes the iteration to terminate at 0 (inclusive).</t>
+                     <t>If <tt>step</tt> is <tt>0</tt>, the resulting array is empty.</t>
+                     <t>If <tt>step</tt> is omitted, it defaults to <tt>1</tt>.</t>
+                     <t>The resulting array is then constructed by iterating from <tt>start</tt> (inclusive)
+                        to <tt>end</tt> (exclusive), collecting array elements at index <tt>i</tt>,
+                        and incrementing the iterator <tt>i</tt> by <tt>step</tt> after each iteration. </t>
+                     <t>If <tt>len</tt> is <tt>0</tt> the resulting array is empty.</t>
+                     <t>If <tt>step &gt; 0</tt> and <tt>start &gt; end</tt> the resulting array is empty.</t>
+                     <t>If <tt>step &lt; 0</tt> and <tt>start &lt; end</tt> the resulting array is empty.</t>
+                  </section>
```

Comment:
I hope these three paragraphs are actually consequences of the preceding text. If so, it would be better to be clear that they are consequences rather than additional rules. E.g. "As a consequence of the above:" follow by a list of three bullets.